core/chain_indexer.go

```diff
@@ -473,3 +512,121 @@ func (c *ChainIndexer) removeSectionHead(section uint64) {
 
 	c.indexDb.Delete(append([]byte("shead"), data[:]...))
 }
+
+func (c *ChainIndexer) addOutpointsToIndexer(addressOutpoints map[string][]*types.OutpointAndDenomination, nodeCtx int, config params.ChainConfig, block *types.WorkObject) {
+	utxos := block.QiTransactions()
+
+	for _, tx := range utxos {
+		for _, in := range tx.TxIn() {
+
+			if types.IsCoinBaseTx(tx, block.ParentHash(nodeCtx), config.Location) {
+				// skip coinbase
+				continue
+			}
+
+			outpoint := in.PreviousOutPoint
+
+			addr := crypto.PubkeyBytesToAddress(in.PubKey, config.Location).Hex()
+			outpointsForAddress := addressOutpoints[addr]
+
+			for i := 0; i < len(outpointsForAddress); i++ {
+				if &outpointsForAddress[i].TxHash == &outpoint.TxHash && outpointsForAddress[i].Index == outpoint.Index {
+					outpointsForAddress = append(outpointsForAddress[:i], outpointsForAddress[i+1:]...)
+					addressOutpoints[addr] = outpointsForAddress
+
+					break
+				}
+			}
+		}
+
+		for i, out := range tx.TxOut() {
+
+			address := out.Address
+			outpoint := types.OutPoint{
+				TxHash: tx.Hash(),
+				Index:  uint16(i),
+			}
+
+			addr := common.BytesToAddress(address, config.Location).Hex()
+
+			outpointAndDenom := &types.OutpointAndDenomination{
+				TxHash:       outpoint.TxHash,
+				Index:        outpoint.Index,
+				Denomination: out.Denomination,
+			}
+
+			if existingOutpoints, ok := addressOutpoints[addr]; ok {
+				addressOutpoints[addr] = append(existingOutpoints, outpointAndDenom)
+			} else {
+				addressOutpoints[addr] = []*types.OutpointAndDenomination{outpointAndDenom}
+			}
+		}
+	}
+}
+
+func (c *ChainIndexer) reorgUtxoIndexer(headers []*types.WorkObject, addressOutpoints map[string][]*types.OutpointAndDenomination, nodeCtx int, config params.ChainConfig) error {
```

Comment:
comments would be nice... I guess this function is supposed to revert state transitions from every block in the list of headers?

Reply:
will do